Use compact proofs for messages delivery

serban300 opened the pull request:
Fixes: #665, #2147

Performance improvement: Almost 50% as measured in #2147 
Small size improvement: for 4096 entries of 2048 bytes each and keys of the form `format!("key{}", NUM_MSGS - i)`
- read proof: 8674561 bytes
- compact proof: 8430549 bytes

serban300 commented on primitives/runtime/src/vec_db.rs:
```diff
@@ -0,0 +1,274 @@
+
+	/// Checks if each key was read.
+	pub fn ensure_no_unused_keys(&self) -> Result<(), VecDbError> {
+		for (key, _) in &self.db {
```
I think that `verify_trie_proof()` does that internally. I mean if we call `generate_trie_proof()` for `{ k1 => v1, k2 => v2 }` and we only `verify_trie_proof()` for `{ k1 => v1 }` I remember that I tried that and it was thorowing an `ExtraneousNode` error.  but I'll double check and come back with an exact answear. And also I'll add a unit test.

svyatonik commented on primitives/runtime/src/vec_db.rs:
```diff
@@ -0,0 +1,274 @@
+use frame_support::{PalletError, StateVersion};
```
Please add file header

svyatonik commented on primitives/runtime/src/vec_db.rs:
```diff
@@ -0,0 +1,274 @@
+		mut keys: Vec<impl AsRef<[u8]> + Ord>,
+	) -> Result<Self, VecDbError> {
+		let mem_db = read_proof.into_memory_db();
+		let trie_db = TrieDBBuilder::<LayoutV1<H>>::new(&mem_db, &root).build();
```
Sorry, I still had no time to explore this layout stuff in details, so probably an absolute dumb q - is it fine that we're always using `LayoutV1` here, but switching between `LayoutV0` and `LayoutV1` when verifying proofs?

svyatonik commented on bin/runtime-common/src/messages_generation.rs:
```diff
@@ -95,13 +98,32 @@ where
 			storage_keys.push(storage_key);
 		}
+
+		if add_duplicate_key {
```
Probably worth add test for [that](https://github.com/paritytech/parity-bridges-common/pull/2155#discussion_r1205446525) here too

svyatonik replied:
I mean test for extra trie node. But NVM - having test in the `vec_db.rs` is enough

svyatonik reviewed (approved):
Looks great, Thank you! Would be ideal to see the impact of this optimization on the whole transaction weight (because proof verification is a (significant) part of the delivery transaction). Also - as you've suggested in the related issue, let's do the same optimization for the confirmation transaction.

serban300 commented:
> Would be ideal to see the impact of this optimization on the whole transaction weight (because proof verification is a (significant) part of the delivery transaction).

Thanks for the suggestions ! Would be actually very good to see this. What would be the best way to do this ? Should I run the benchmarks ?

svyatonik commented:
> Thanks for the suggestions ! Would be actually very good to see this. What would be the best way to do this ? Should I run the benchmarks ?

I think anyone could do that, even post-merge. Just run the same benchmarks (messages pallet) on new version and on the parent commit. So it isn't a blocker for merge - just something that would be interesting for the future us (how much did we gain with this optimization) :)

serban300 commented:
> > Thanks for the suggestions ! Would be actually very good to see this. What would be the best way to do this ? Should I run the benchmarks ?
> 
> I think anyone could do that, even post-merge. Just run the same benchmarks (messages pallet) on new version and on the parent commit. So it isn't a blocker for merge - just something that would be interesting for the future us (how much did we gain with this optimization) :)

Better to run the benchmarks before merging. Trying to do that now, but I'm getting some `InvalidMessagesProof` error. Will push new commits after I sort this out.

serban300 commented:
I ran the benchmarks. I don't see big differences, thay are probably even withing the error margin. But from what I understand we only measure scenarios in which we send 1 or 2 messages and apart from that the map is clean. The differences should be noticeable with more messages and bigger maps.

@svyatonik wdyt ?

serban300 commented:
Did another test to get more relevant performance results. I modified `receive_single_message_proof` to actually send 1001 messages instead of just 1 and also removed the error logs that can affect the performance:

```
	receive_single_message_proof {
		let relayer_id_on_source = T::bridged_relayer_id();
		let relayer_id_on_target = account("relayer", 0, SEED);
		T::endow_account(&relayer_id_on_target);

		// mark messages 1..=20 as delivered
		receive_messages::<T, I>(20);

		let (proof, dispatch_weight) = T::prepare_message_proof(MessageProofParams {
			lane: T::bench_lane_id(),
			message_nonces: 21..=1021,
			outbound_lane_data: None,
			is_successful_dispatch_expected: false,
			size: StorageProofSize::Minimal(EXPECTED_DEFAULT_MESSAGE_LENGTH),
		});
	}: receive_messages_proof(RawOrigin::Signed(relayer_id_on_target), relayer_id_on_source, proof, 1001, dispatch_weight)
	verify {
		assert_eq!(
			crate::InboundLanes::<T, I>::get(&T::bench_lane_id()).last_delivered_nonce(),
			1021,
		);
	}
```

The results are:

With the old code:

```
	fn receive_single_message_proof() -> Weight {
		// Proof Size summary in bytes:
		//  Measured:  `490`
		//  Estimated: `52645`
		// Minimum execution time: 10_002_807 nanoseconds.
		Weight::from_parts(10_105_449_000, 52645)
			.saturating_add(T::DbWeight::get().reads(3_u64))
			.saturating_add(T::DbWeight::get().writes(1_u64))
	}
``` 

With the new code:
```
	fn receive_single_message_proof() -> Weight {
		// Proof Size summary in bytes:
		//  Measured:  `490`
		//  Estimated: `52645`
		// Minimum execution time: 5_307_999 nanoseconds.
		Weight::from_parts(5_356_684_000, 52645)
			.saturating_add(T::DbWeight::get().reads(3_u64))
			.saturating_add(T::DbWeight::get().writes(1_u64))
	}
```

So the improvement is indeed significant (about 50%). But it's hard to notice when sending only one message.

acatangiu commented:
> So the improvement is indeed significant (about 50%). But it's hard to notice when receiving only one message.

1001 messages is also I think unrealistic for a single proof, but the results are very encouraging :100: 

If cost is mostly the same for 1 message, but only goes down (in comparison with old version) as more messages are included, then this definitely a clean win! :tada:

svyatonik commented:
So the good thing is that Serban has improved performance (thank you). The bad thing is that we don't see this improvement - call weight will be close to the previous value. So IMO we need a follow up PR with benchmarks update - we need some better estimation than we have now (roughly speaking: call weight = `NumberOfMessages * (WeightOfTwoMessages - WeightOfOneMessage)`). Probably instead of `1` and `2` messages use the `1..=MAX` interval - needs to be investigated.

That said, imo this PR is good to go.

serban300 commented:
> So IMO we need a follow up PR with benchmarks update - we need some better estimation than we have now (roughly speaking: call weight = `NumberOfMessages * (WeightOfTwoMessages - WeightOfOneMessage)`). Probably instead of `1` and `2` messages use the `1..=MAX` interval - needs to be investigated.

Agree that it would be good to see if we can have some better estimation for the benchmarks. I will create an issue for this and do some research.

acatangiu reviewed (approved):
Nice!

bkontur commented on primitives/runtime/src/vec_db.rs:
```diff
@@ -0,0 +1,332 @@
+
+impl UntrustedVecDb {
+	/// Creates a new instance of `UntrustedVecDb`.
+	pub fn new<H: Hasher>(
```
small nit, `new -> try_new`

bkontur commented on primitives/runtime/src/vec_db.rs:
```diff
@@ -0,0 +1,332 @@
+		let trie_proof = generate_trie_proof::<LayoutV1<H>, _, _, _>(&mem_db, root, &keys)
+			.map_err(|_| VecDbError::UnableToGenerateTrieProof)?;
+
+		let mut entries = vec![];
```
Couldnt we initialize `Vec::with_capacity(keys.len())` or something like that?

serban300 replied:
Yes. Missed this. Thanks ! Done.
